go/consensus/tendermint/abci/mux.go

```diff
@@ -411,16 +403,48 @@ func (mux *abciMux) InitChain(req types.RequestInitChain) types.ResponseInitChai
 	}
 
 	// Dispatch registered genesis hooks.
-	mux.RLock()
-	defer mux.RUnlock()
+	func() {
+		mux.RLock()
+		defer mux.RUnlock()
+
+		mux.logger.Debug("Dispatching genesis hooks")
+
+		for _, hook := range mux.genesisHooks {
+			hook()
+		}
+
+		mux.logger.Debug("Genesis hook dispatch complete")
+	}()
+
+	// TODO: remove stateKeyGenesisRequest here, see oasis-core#2426
+	b, _ = req.Marshal()
+	mux.state.deliverTxTree.Set([]byte(stateKeyGenesisRequest), b)
+	mux.state.checkTxTree.Set([]byte(stateKeyGenesisRequest), b)
+
+	// Call InitChain() on all applications.
+	mux.logger.Debug("InitChain: initializing applications")
+	ctx := NewContext(ContextInitChain, mux.currentTime, mux.state)
 
-	mux.logger.Debug("Dispatching genesis hooks")
+	for _, app := range mux.appsByLexOrder {
+		mux.logger.Debug("InitChain: calling InitChain on application",
+			"app", app.Name(),
+		)
 
-	for _, hook := range mux.genesisHooks {
-		hook()
+		if err = app.InitChain(ctx, req, st); err != nil {
+			mux.logger.Error("InitChain: fatal error in application",
+				"err", err,
+				"app", app.Name(),
+			)
+			panic("mux: InitChain: fatal error in application: '" + app.Name() + "': " + err.Error())
+		}
 	}
 
-	mux.logger.Debug("Genesis hook dispatch complete")
+	mux.logger.Debug("InitChain: initializing of applications complete", "num_collected_events", len(ctx.GetEvents()))
+
+	// Since returning emitted events doesn't work for InitChain() response yet,
+	// we store those and return them in BeginBlock().
+	evJSON, _ := json.Marshal(ctx.GetEvents())
```

Comment:
Oh didn't notice this before, but why do you use JSON to serialize this? Replace it with CBOR serialization.

Reply:
No reason actually.. I used it for debugging and then left it there. :/